File: leetcode/string/Q125/Solution.java
package leetcode.string.Q125;

import java.util.Scanner;

public class Solution {
    public static boolean isPalindrome(String s) {
        if (s.isBlank()) {
            return true;
        }

        String regex = "[^a-z0-9]";
        s = s.toLowerCase()
                .replaceAll(regex, "");
        int end = s.length();

        for (int i = 0; i < end / 2; i++) {
            if (s.charAt(i) != s.charAt(end - i - 1)) {
                return false;
            }
        }

        return true;
    }
}

File: leetcode/string/q125/Solution.kt
package leetcode.string.q125

import java.util.*

fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    val s = sc.nextLine()

    sc.close()

    val palindrome = isPalindrome(s)
    println(palindrome)
}

fun isPalindrome(s: String): Boolean {
    var start = 0
    var end = s.length - 1

    while (start < end) {
        when {
            !Character.isLetterOrDigit(s[start]) -> start++
            !Character.isLetterOrDigit(s[end]) -> end--
            Character.toLowerCase(s[start]) != Character.toLowerCase(s[end]) -> return false
            else -> {
                start++
                end--
            }
        }
    }

    return true
}
